fix(components): Fix form not always focusing first error

MichaelParadis opened the pull request:
<!--
  Atlantis uses Conventional Commits to track versions.
  Pull request titles should follow the following format.

  For help creating your pull request, you can [use this tool](https://atlantis.getjobber.com/?path=/story/guides-pull-request-title-generator--docs)

  <TYPE>(<optional SCOPE>): <conditionally BREAKING CHANGE:> <description>

  eg.
    fix(SCOPE): stop graphite breaking when too much pressure applied — Patch Release
    feat(SCOPE): add 'graphiteWidth' option — (Minor) Feature Release
    feat(SCOPE): BREAKING CHANGE: remove graphiteWidth option — (Major) Breaking Release

  TYPE should consist of:
    - fix: a commit of the type fix patches a bug in your codebase
    - feat: a commit of the type feat introduces a new feature to the codebase
    - docs: documentation only changes
    - build: improvements to the build system
    - refactor: a change that neither fixes a bug nor introduces a feature
    - chore: other changes that don't modify src or test files

  SCOPE should be one of:
    - components
    - components-native
    - deps
    - deps-dev
    - design
    - docx
    - eslint
    - formatters
    - generators
    - hooks
    - stylelint


  If your pull request introduces a breaking change please append `BREAKING CHANGE:` following type / scope.

  Further Reading:
    - https://www.conventionalcommits.org
    - https://github.com/commitizen/conventional-commit-types/blob/master/index.json
-->

## Motivations

<!-- Why did you do what you did? -->

## Changes

<!-- https://keepachangelog.com/en/1.0.0/ -->

### Added

- <!-- new features -->

### Changed

- <!-- changes in existing functionality -->

### Deprecated

- <!-- soon-to-be removed features -->

### Removed

- <!-- now removed features -->

### Fixed

- <!-- for any bug fixes -->

### Security

- <!-- in case of vulnerabilities -->

## Testing

<!-- How to test your changes. -->

Changes can be
[tested via Pre-release](https://github.com/GetJobber/atlantis/actions/workflows/trigger-qa-build.yml)

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

![Random photo of Atlantis](https://loremflickr.com/672/400/atlantis)

ericchernuka commented on packages/components/src/Form/Form.tsx:
```diff
@@ -86,11 +87,8 @@ export const Form = forwardRef(function InternalForm(
-    const element = document.querySelector(
-      `[name="${firstErrName}"]`,
-    ) as HTMLElement;
+    const errorKeys = Object.keys(errs);
 
-    element?.focus();
+    setFocus(errorKeys[0]);
   }
 });
```
Any reason we couldn't leverage the built-in functionality from RHF? https://react-hook-form.com/docs/useform#shouldFocusError

MichaelParadis replied:
I was trying that and it didn't seem to work as expected. I can try again and see if it was a weird caching thing

ericchernuka replied:
Wonder if it's an issue with the register note 🤔

MichaelParadis replied:
It turns out that after I added the refs in FormField it worked as expected. I assume it had to do with 
```
Only registered fields with a ref will work. 
```

and a ref wasn't registered until my change

ericchernuka replied:
I haven't checked the code, but looks like the use controller ref would need to be passed to the component? https://react-hook-form.com/docs/usecontroller

ericchernuka replied:
Nice!

ericchernuka commented on packages/components/src/FormField/FormField.tsx:
```diff
@@ -148,7 +146,14 @@ function FormFieldInternal(props: FormFieldInternalProps) {
             {...textFieldProps}
             rows={rows}
-            ref={inputRef as MutableRefObject<HTMLTextAreaElement>}
+            ref={ref => {
+              if (inputRef) {
+                (
+                  inputRef as MutableRefObject<HTMLTextAreaElement | null>
+                ).current = ref;
+              }
+              rest.ref(ref);
+            }}
```
In the past I've leveraged merging refs using a util.
```typescript
export function mergeRefs<T = any>(
  refs: Array<React.MutableRefObject<T> | React.LegacyRef<T> | undefined | null>
): React.RefCallback<T> {
  return (value) => {
    refs.forEach((ref) => {
      if (typeof ref === "function") {
        ref(value);
      } else if (ref != null) {
        (ref as React.MutableRefObject<T | null>).current = value;
      }
    });
  };
}
```

MichaelParadis commented on packages/components/src/FormField/FormField.tsx:
```diff
@@ -119,6 +122,7 @@ function FormFieldInternal(props: FormFieldInternalProps) {
     autoFocus: autofocus,
     onKeyDown: handleKeyDown,
   };
+  const inputRefs = mergeRefs([inputRef, fieldRef]);
```
Uses the field ref from `useController` like we do in components-native [here](https://github.com/GetJobber/atlantis/blob/master/packages/components-native/src/InputText/InputText.tsx#L435-L438)

MichaelParadis commented on packages/components/src/FormField/FormField.tsx:
```diff
@@ -58,12 +58,10 @@ function FormFieldInternal(props: FormFieldInternalProps) {
-      ? useFormContext()
-      : // If there isn't a Form Context being provided, get a form for this field.
-        useForm({ mode: "onTouched" });
+    formContext ?? useForm({ mode: "onTouched" });
```
`shouldFocus` is the default on `useForm` https://react-hook-form.com/docs/useform#shouldFocusError

MichaelParadis replied:
Yes, for some reason I thought you were referencing the trigger with shouldFocus which is needed because trigger doesn't focus by default even with shouldFocusError being true.
Pushing the change now

MichaelParadis commented on packages/components/src/Form/Form.test.tsx:
```diff
@@ -152,14 +153,20 @@ it("should focus on the first errored field", async () => {
+    expect(inputTwo).toHaveFocus();
+  });
+
+  userEvent.click(getByText("submit"));
```
Good call on updating the test suite. I will get right on that

MichaelParadis commented:
@jdeichert I updated the Form tests in [0e4a246](https://github.com/GetJobber/atlantis/pull/1970/commits/0e4a246478edae68188d05806ee00792b3c3fb27). I also renamed all instances of `test("...")` to `it(...` to match our other files in [28a850b](https://github.com/GetJobber/atlantis/pull/1970/commits/28a850b5d9b7afa733b3f83ead600247169a6039)

I added a comment and some tests for mergeRefs in [705f98a](https://github.com/GetJobber/atlantis/pull/1970/commits/705f98a4d4742e7c1fde5028b71c3651333e9785)
